sklearn/metrics/tests/test_classification.py

```diff
@@ -483,6 +483,35 @@ def test_matthews_corrcoef_multiclass():
     assert_almost_equal(mcc, 0.)
 
 
+def test_matthews_corrcoef_overflow():  # Bug#9622
+    rng = np.random.RandomState(20170906)
+
+    def mcc_safe(y_true, y_pred):
+        conf_matrix = confusion_matrix(y_true, y_pred)
+        true_pos = conf_matrix[1,1]
+        false_pos = conf_matrix[1,0]
+        false_neg = conf_matrix[0,1]
+        n_points = len(y_true)
+        pos_rate = (true_pos + false_neg) / n_points
+        activity = (true_pos + false_pos) / n_points
+        mcc_numerator = true_pos / n_points - pos_rate * activity
+        mcc_denominator = activity * pos_rate * (1 - activity) * (1 - pos_rate)
+        return mcc_numerator / np.sqrt(mcc_denominator)
+
+    def random_ys(n_points):
+        x_true = rng.random_sample(n_points)
+        x_pred = x_true + 0.2 * (rng.random_sample(n_points) - 0.5)
+        y_true = (x_true > 0.5)
+        y_pred = (x_pred > 0.5)
+        return y_true, y_pred
+
+    for n_points in [100, 10000, 1000000]:
+        y_true, y_pred = random_ys(n_points)
+        assert_almost_equal(matthews_corrcoef(y_true, y_true), 1.0)
```

Comment:
Just checking this with a simple array is a perfectly fine regression test for the bug we are seeing. A simpler example showing the bug:

```py
import numpy as np
from sklearn.metrics.classification import matthews_corrcoef


n = int(1e5)
arr = np.repeat([0., 1.], n)
print(matthews_corrcoef(arr, arr))

# multiclass test for good measure
arr = np.repeat([0., 1., 2.], n)
print(matthews_corrcoef(arr, arr))
```

Reply:
given the lack of transparent bignums in numpy, your magic is dangerous.
I don't mind the other change ("manipulating quantities that are of order 1"), but casting to `float` is critical.

Reply:
I will gladly add the simple test, but I see no reason to give up the big random one.